File: crackingthecodeinterview/trees/adjlist.cpp
#include <iostream>
#include <vector>
#include <queue>

using std::ostream;
using std::cout;
using std::cerr;
using std::vector;
using std::queue;
using std::endl;

struct Vertex{
    char            vertexchar;
    char            color;
    int             distance;
    Vertex*         predessesor; 
    vector<Vertex*> verticies;

    Vertex(void) : vertexchar('a'), color('w'), distance(-1), predessesor(NULL) {}
    Vertex(char vertchar) : vertexchar(vertchar), color('w'), distance(-1), predessesor(NULL) {}
    Vertex(char vertchar, int distance) : vertexchar(vertchar), color('w'), distance(distance), predessesor(NULL) {}

    friend ostream &operator<<(ostream& os, const Vertex& in)
    {
        os << in.vertexchar << "-" << in.color << "-" << in.predessesor << " ";
        return os;
    }
};

struct Adjlist{
    vector<Vertex*> list;

    friend ostream &operator<<(ostream& os, const Adjlist& in)
    {
        for(int i = 0; i < in.list.size(); i++){
            os << in.list[i]->vertexchar << ": ";
            for(int j = 0; j < in.list[i]->verticies.size(); j++)
                os << *in.list[i]->verticies[j];
            os << endl;
        }
        return os;
    }
};

void
BFS(Vertex* v)
{
    queue<Vertex*> inqueue;
    Vertex*        tempvertex;
    int            i = 0;

    v->color    = 'g';
    v->distance = 0;  

    inqueue.emplace(v);
    while(!inqueue.empty()){
        tempvertex = inqueue.front();
        inqueue.pop();
        //need to add iterator here.. to modify pointers
        for( Vertex* k : tempvertex->verticies )
            if(k->color == 'w'){
                k->color       = 'g';
                k->distance    = tempvertex->distance + 1;
                k->predessesor = tempvertex; 
                inqueue.emplace(k);
            }
        v->color = 'b';
    }
} 
            

int
main()
{
    Adjlist temp;
    Vertex* a = new Vertex('a');
    Vertex* b = new Vertex('b');
    Vertex* c = new Vertex('c');
    Vertex* d = new Vertex('d');
    Vertex* e = new Vertex('e');
    Vertex* f = new Vertex('f');
    Vertex* g = new Vertex('g');
    temp.list.push_back(a);
    temp.list[0]->verticies.push_back(b);
    temp.list.push_back(b);
    temp.list[1]->verticies.push_back(c);
    temp.list[1]->verticies.push_back(d);
    temp.list.push_back(c);
    temp.list[2]->verticies.push_back(d);
    temp.list[2]->verticies.push_back(e);
    temp.list.emplace_back(d);
    temp.list[3]->verticies.push_back(e);
    temp.list[3]->verticies.push_back(a);
    temp.list.push_back(e);
    temp.list[4]->verticies.push_back(f);
    temp.list.push_back(f);
    temp.list.push_back(g);
    temp.list[6]->verticies.emplace_back(d); 

    Vertex* to   = a;
    Vertex* from = f;
    bool    swch  = false;
    BFS(to);

    while(from != NULL){
        if(from == to)
            swch = true;
        from = from->predessesor;
    }
    cout <<  swch << endl;
    return 0;
}

File: crackingthecodeinterview/trees/adjlist.py
class Vertex:
    
    def __init__(self, char=''):
        self.color    = 'w'
        self.firstime = -1
        self.lastime  = -1
        self.char     = char
        self.prevVert = None
        self.distance = 0
        self.adjvert  = []

    def __str__(self):
        return ("Vertex(" + self.char + "): " + str(self.distance) + '-' + self.color + "-(" + str(self.firstime) + ")-(" + str(self.lastime) + ")")

class AdjList:
    
    def __init__(self):
        self.list = []

    def __str__(self):
        string = ""
        for i in self.list:
            string += (str(i))
            for j in i.adjvert:
                string += ("->" + str(j))
            string += "\n"
        return string

def BFS(v):
    inqueue = []
    tempver = Vertex()
    
    v.color    = 'g'
    v.distance = 0

    inqueue.append(v)
    while(len(inqueue) != 0):
        tempver = inqueue.pop()

        for i in tempver.adjvert:
            if i.color == 'w':
                i.color = 'g'
                i.distance = tempver.distance + 1
                i.prevVert = tempver
                inqueue.append(i)
        v.color = 'b'

def main():
    temp = AdjList()
    
    a = Vertex('a')
    b = Vertex('b')
    c = Vertex('c')
    d = Vertex('d')
    e = Vertex('e')
    f = Vertex('f')
    g = Vertex('g')

    temp.list.append(a);
    temp.list[0].adjvert.append(b);
    temp.list.append(b);
    temp.list[1].adjvert.append(c);
    temp.list[1].adjvert.append(d);
    temp.list.append(c);
    temp.list[2].adjvert.append(d);
    temp.list[2].adjvert.append(e);
    temp.list.append(d);
    temp.list[3].adjvert.append(a);
    temp.list[3].adjvert.append(e);
    temp.list.append(e);
    temp.list[4].adjvert.append(f);
    temp.list.append(f);
    temp.list.append(g);
    temp.list[6].adjvert.append(d); 

    BFS(a)

    print(temp)
    
    vert = e 
    while(vert.prevVert != None):
        vert = vert.prevVert

    print(vert)

main()
